tpm2/src/util/nsl_util.c

```diff
@@ -0,0 +1,983 @@
+#include <openssl/evp.h>
+#include <openssl/aes.h>
+
+#include <openssl/rand.h>
+#include <openssl/rsa.h>
+#include <openssl/bn.h>
+#include <openssl/engine.h>
+
+#include "defines.h"
+
+//
+// encrypt_with_key_pair()
+//
+int encrypt_with_key_pair(EVP_PKEY_CTX * ctx,
+                          const unsigned char *p, size_t p_len,
+                          unsigned char **c, size_t *c_len)
+{
+  // Initialize the context for encryption.
+  int result = EVP_PKEY_encrypt_init(ctx);
+
+  if (result == 0)
+  {
+    kmyth_log(LOG_ERR, "Failed to initialize the EVP context for encryption.");
+    return 1;
+  }
+
+  // Determine the length of the ciphertext buffer.
+  result = EVP_PKEY_encrypt(ctx, NULL, c_len, p, p_len);
+  if (result <= 0)
+  {
+    kmyth_log(LOG_ERR,
+              "Failed to determine the length of the ciphertext buffer.");
+    return 1;
+  }
+
+  // Allocate the ciphertext buffer.
+  *c = calloc(*c_len, sizeof(unsigned char));
+  if (*c == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the ciphertext buffer.");
+    return 1;
+  }
+
+  // Encrypt the plaintext.
+  result = EVP_PKEY_encrypt(ctx, *c, c_len, p, p_len);
+  if (result <= 0)
+  {
+    kmyth_log(LOG_ERR, "Failed to encrypt the plaintext.");
+    return 1;
+  }
+
+  return 0;
+}
+
+//
+// decrypt_with_key_pair()
+//
+int decrypt_with_key_pair(EVP_PKEY_CTX * ctx,
+                          const unsigned char *c, size_t c_len,
+                          unsigned char **p, size_t *p_len)
+{
+  // Initialize the context for decryption.
+  int result = EVP_PKEY_decrypt_init(ctx);
+
+  if (result == 0)
+  {
+    kmyth_log(LOG_ERR, "Failed to initialize the EVP context for decryption.");
+    return 1;
+  }
+
+  // Determine the length of the plaintext buffer.
+  result = EVP_PKEY_decrypt(ctx, NULL, p_len, c, c_len);
+  if (result <= 0)
+  {
+    kmyth_log(LOG_ERR,
+              "Failed to determine the length of the plaintext buffer.");
+    return 1;
+  }
+
+  // Allocate the plaintext buffer.
+  *p = calloc(*p_len, sizeof(unsigned char));
+  if (*p == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the plaintext buffer.");
+    return 1;
+  }
+
+  // Decrypt the ciphertext.
+  result = EVP_PKEY_decrypt(ctx, *p, p_len, c, c_len);
+  if (result <= 0)
+  {
+    kmyth_log(LOG_ERR, "Failed to decrypt the ciphertext.");
+    return 1;
+  }
+
+  return 0;
+}
+
+//
+// build_nonce_request()
+// 
+int build_nonce_request(EVP_PKEY_CTX * ctx,
+                        unsigned char *nonce, size_t nonce_len,
+                        unsigned char *id, size_t id_len,
+                        unsigned char **request, size_t *request_len)
+{
+  // Allocate the unencrypted request buffer.
+  unsigned char *message = NULL;
+  size_t message_len = id_len + nonce_len + (2 * sizeof(size_t));
+
+  message = calloc(message_len, sizeof(unsigned char));
+  if (message == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the message buffer.");
+    return 1;
+  }
+
```

Comment:
Should we validate the nonce and id length fields when constructing the message or just leave that check to the recipient (and the calling function?)?

Reply:
I've added a check for the nonce length like the other checks. For now I'll leave the ID length check as `TODO`.